File: calculations.py
import math
class calculations:
    def WCA(self, windDirection, windSpeed, trueCourse, trueAirSpeed):
            crossWindAngle = windSpeed * math.sin(math.radians(windDirection - trueCourse))
            headwindAngle = math.degrees(windSpeed * math.cos(math.radians(windDirection - trueCourse)))
            windCorrectionAngle = math.degrees(math.asin(crossWindAngle / trueAirSpeed))
            groundSpeed = trueAirSpeed - headwindAngle
            trueHeading = self.confirCompassHeading(trueCourse + windCorrectionAngle)
            return {"WCA":round(windCorrectionAngle), "TH":round(trueHeading), "GS":round(groundSpeed)}
        
    def compassHeading(self, magneticHeading, compassDeviation):
        compassHeading = self.confirCompassHeading(magneticHeading + compassDeviation)
        return round(compassHeading)
        
    def magneticHeading(self, trueHeading, variation):
        magneticHeading = self.confirCompassHeading(trueHeading + variation)
        return round(magneticHeading)
        
    def confirCompassHeading(self, heading):
        if heading < 0:
            heading += 360
        elif heading >= 360:
            heading -= 360
        return heading
        
    def timeEnroute(self, distance, groundSpeed):
        timeEnroute = distance / groundSpeed
        timeEnrouteHours = timeEnroute
        hours = int(timeEnroute)
        minutes = (timeEnroute - hours) * 60
        seconds = (minutes - int(minutes)) * 60
        minutes = minutes + (hours * 60)
        return {"time":f"{int(minutes)}:{int(seconds):02d}", "hrs":timeEnrouteHours}
        
    def fuelUsed(self, fuelFlow, timeEnrouteHours):
        fuelUsed = fuelFlow * timeEnrouteHours
        return round(fuelUsed, 2)

File: react-app/src/js/calculations.js
    function WCA(windDirection, windSpeed, trueCourse, trueAirSpeed) {
        const crossWindAngle = windSpeed * Math.sin(radians(windDirection - trueCourse));
        console.log(crossWindAngle);
        const windCorrectionAngle = degrees(Math.asin(crossWindAngle / trueAirSpeed));
        const groundSpeed = Math.sqrt(
        trueAirSpeed ** 2 + windSpeed ** 2 -
        2 * trueAirSpeed * windSpeed * Math.cos(radians(trueCourse - windDirection + windCorrectionAngle))
        );
        const trueHeading = confirCompassHeading(trueCourse + windCorrectionAngle);

        return {
            WCA: Math.round(windCorrectionAngle),
            TH: Math.round(trueHeading),
            GS: Math.round(groundSpeed)
        };
    }
        
    function compassHeading(magneticHeading, compassDeviation) {
        let compassHeading = confirCompassHeading(magneticHeading + compassDeviation)
        return Math.round(compassHeading)
    }

    function magneticHeading(trueHeading, variation){
        let magneticHeading = confirCompassHeading(trueHeading + variation)
        return Math.round(magneticHeading)
    }
        
    function confirCompassHeading(heading){
        if (heading < 0) {
            heading += 360
            return heading
        }
        if (heading >= 360) {
            return heading -= heading - 360
        }
        return heading
    }
        
    function timeEnroute(distance, groundSpeed){
        let timeEnroute = distance / groundSpeed
        let timeEnrouteHours = timeEnroute
        let hours = Math.floor(timeEnroute)
        let minutes = (timeEnroute - hours) * 60
        let seconds = (minutes - Math.floor(minutes)) * 60
        minutes = minutes + (hours * 60)
        return {"time":`${Math.floor(minutes)}:${Math.floor(seconds).toString().padStart(2, '0')}`, "hrs":timeEnrouteHours}
    }
        
    function fuelUsed(fuelFlow, timeEnrouteHours){
        let fuelUsed = fuelFlow * timeEnrouteHours
        return (Math.round(fuelUsed * 100) / 100)
    }

    function radians(degrees) {
        return (degrees * (Math.PI / 180));
    }

    function degrees(radians) {
        return (radians * (180 / Math.PI));
    }

    export { WCA, compassHeading, magneticHeading, timeEnroute, fuelUsed };